src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -189,6 +130,90 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
             }
         }
 
+        private void HandlePublicProperties()
+        {
+            PropertyInfo[] properties = Type.GetProperties(BindingFlags.Instance | BindingFlags.Public);
+
+            Dictionary<string, JsonPropertyInfo> cache = CreatePropertyCache(properties.Length);
+
+            foreach (PropertyInfo propertyInfo in properties)
+            {
+                // Ignore indexers
+                if (propertyInfo.GetIndexParameters().Length > 0)
+                {
+                    continue;
+                }
+
+                // For now we only support public getters\setters
+                if (propertyInfo.GetMethod?.IsPublic == true ||
+                    propertyInfo.SetMethod?.IsPublic == true)
+                {
+                    JsonPropertyInfo jsonPropertyInfo = AddProperty(
+                        propertyInfo,
+                        Type,
+                        Options);
+                    Debug.Assert(jsonPropertyInfo != null && jsonPropertyInfo.NameAsString != null);
+
+                    // If the JsonPropertyNameAttribute or naming policy results in collisions, throw an exception.
+                    if (!JsonHelpers.TryAdd(cache, jsonPropertyInfo.NameAsString, jsonPropertyInfo))
+                    {
+                        JsonPropertyInfo other = cache[jsonPropertyInfo.NameAsString];
+
+                        if (other.ShouldDeserialize == false && other.ShouldSerialize == false)
+                        {
+                            // Overwrite the one just added since it has [JsonIgnore].
+                            cache[jsonPropertyInfo.NameAsString] = jsonPropertyInfo;
+                        }
+                        else if (jsonPropertyInfo.ShouldDeserialize == true || jsonPropertyInfo.ShouldSerialize == true)
+                        {
+                            ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(Type, jsonPropertyInfo);
+                        }
+                        // else ignore jsonPropertyInfo since it has [JsonIgnore].
+                    }
+                }
+            }
+
+            JsonPropertyInfo[] cacheArray;
+            if (DetermineExtensionDataProperty(cache))
+            {
+                // Remove from cache since it is handled independently.
+                cache.Remove(DataExtensionProperty!.NameAsString!);
+
+                cacheArray = new JsonPropertyInfo[cache.Count + 1];
+
+                // Set the last element to the extension property.
+                cacheArray[cache.Count] = DataExtensionProperty;
+            }
+            else
+            {
+                cacheArray = new JsonPropertyInfo[cache.Count];
+            }
+
+            // Set fields when finished to avoid concurrency issues.
+            PropertyCache = cache;
+            cache.Values.CopyTo(cacheArray, 0);
+            PropertyCacheArray = cacheArray;
+        }
+
+        private void HandleNonPublicProperties()
```

Comment:
This should be able to be combined with the HandlePublicProperties

Reply:
Do you mean can combine the logic into one method...

or do you mean we can include the `BindingFlags.NonPublic` flag in the `GetProperties` call and do some if-check to see if the PropertyInfo is non-public? It doesn't seem like there's any property on `PropertyInfo` that can tell us if the property is non-public...

Reply:
(Pushed a commit to minimize the diff in the ctor by inlining the non-public property/`JsonInclude` validation logic).

Reply:
Looks like we can use `IsPublic` on the getters/setters:

```csharp
private static bool IsNonPublicProperty(PropertyInfo propertyInfo)
{
    MethodInfo? getMethod = propertyInfo.GetMethod;
    MethodInfo? setMethod = propertyInfo.SetMethod;
    return !((getMethod != null && getMethod.IsPublic) || (setMethod != null && setMethod.IsPublic));
}
```

Will push a commit